Add a warning when run as root (e.g., sudo pip)

winsonluk opened the pull request:
This change discourages the usage of pip as root by showing a warning when pip install is run with sudo.

```
$ sudo pip3 install
WARNING: Running pip as root is not recommended. This could result in the corruption of the package directories when a package manager is used. See #6409 for more details.
ERROR: You must give at least one requirement to install (see "pip help install")
```

Fixes #6409 
...and potentially helps helps prevent hundreds of future issues related to `sudo pip`: https://github.com/pypa/pip/issues?q=is%3Aissue+%22sudo+pip%22

hroncok commented:
The Fedora's patch: https://src.fedoraproject.org/rpms/python-pip/blob/master/f/emit-a-warning-when-running-with-root-privileges.patch

winsonluk commented:
> The Fedora's patch: https://src.fedoraproject.org/rpms/python-pip/blob/master/f/emit-a-warning-when-running-with-root-privileges.patch

that's some good stuff, thanks. i added the venv check and updated the warning message with the tip

hroncok commented on src/pip/_internal/commands/install.py:
```diff
@@ -226,6 +228,31 @@ def add_options(self):
+                            "This could result in the corruption of the "
+                            "package directories when a package manager is "
+                            "used. Try `%s install --user` instead. "
+                            "See #6409 for the reason behind this warning."
```
I am not sure all users will understand where to look for `#6409` or what it stands for.

winsonluk replied:
good point - removed

uranusjr commented on src/pip/_internal/commands/install.py:
```diff
@@ -1,9 +1,11 @@
+import ctypes
```
`ctypes` is an optional module (only available in CPython when compiled with libffi), so this needs a fallback. I think it’s reasonable to simply disable the superuser check on Windows if ctypes is unavailable.

winsonluk replied:
thanks for the flag. we can catch the import with the existing try-except:

```
             if sys.platform == 'darwin' or sys.platform == 'linux':
                 is_sudo = os.getuid() == 0
             elif sys.platform == 'win32' or sys.platform == 'cygwin':
+                import ctypes
                 is_sudo = ctypes.windll.shell32.IsUserAnAdmin() != 0
-        except AttributeError:
+        except (ImportError, AttributeError):
             is_sudo = False
```

uranusjr commented on src/pip/_internal/commands/install.py:
```diff
@@ -226,6 +228,30 @@ def add_options(self):
+
+        is_venv = (hasattr(sys, 'real_prefix')
+                   or (hasattr(sys, 'base_prefix')
+                       and sys.base_prefix != sys.prefix))
```
https://github.com/pypa/pip/blob/7369ac2ee85815d4a059c667cc415f2000df7ed5/src/pip/_internal/utils/virtualenv.py#L37-L41

winsonluk replied:
Changed to `if is_sudo and not running_under_virtualenv():`

uranusjr commented on src/pip/_internal/commands/install.py:
```diff
@@ -226,6 +228,30 @@ def add_options(self):
+                            "This could result in the corruption of the "
+                            "package directories when a package manager is "
+                            "used. Try `%s install --user` instead."
+                            % command)
```
It’s probably a good idea to use different messages on Windows and POSIX, since “root” does not mean anything on Windows (and the check was not looking for root in the first place).

The final “Try ...” message is problematic as well since it is almost impossible to display a perfect message here, see https://github.com/pypa/pip/pull/7532#discussion_r362325906. We should try our best avoid showing a comment if possible. I would simply say something like “Consider using --user instead”.

Finally, the same check should be applied to `uninstall` as well IMO (without a recommendation, just warn the user this may break the system unless you are sure what you’re doing).

pradyunsg replied:
> Consider using --user instead

This is unnecessary, given the fix for #1668 automatically falls back to the user folders if the global location is unwritable.

winsonluk replied:
Sounds good - we can substitute "root" for "Administrator" for Windows, and take out `Try %s install --user`

I'll update this PR on Monday

pfmoore replied:
Actually, on Windows, running pip as administrator is fine, as there is no "system managed" Python package anyway - so installing via pip as administrator is the correct way of updating your system environment.

So I'd suggest restricting this warning to Unix systems only.

winsonluk replied:
Made some changes:
- Added `Consider using --user instead.` from @uranusjr 's suggestion
- Added this sudo check to uninstall from @uranusjr 's suggestion
- Disabled this check from Windows from @pfmoore 's suggestion

pradyunsg commented:
/cc @dstufft since last I checked he was opposed to just doing this.

pradyunsg commented:
There's still one sticking point with this: docker containers give you root by default.

I don't think we should be calling out the use of pip in root shells like that.

hroncok commented:
I think we should. Even in docker, installing with pip as root can lead to serious problems.

pradyunsg commented:
Perhaps. But that's still more disruptive than I'd like.

I haven't seen a Dockerfile that's using pip that even tries to install as non-root. Is it even possible to use it as non-root from a Dockerfile?

hroncok commented:
It is possible, however not that straightforward. I get your point.

uranusjr commented:
Is this really that disruptive? It’s “just” a warning, pip ignore warnings all the time when they feel like it. Using root in a Docker container is ultimately still not best practice, and it’s arguably still not a bad idea to show the warning in this situation.

If it’s really that important, we can have `--disable-root-check` for people to set if they think they know what they’re doing.

hroncok commented:
`--root /` is for people who think they know they are doing ;)

winsonluk commented:
Yeah, most people don't use Docker, so I think the benefit of preventing mis-installations exceeds whatever disruption a one-line message causes to Docker users.

Plus this is just a `logger.info`, not a `logger.warn`.

xavfernandez commented:
> Perhaps. But that's still more disruptive than I'd like.
> 
> I haven't seen a Dockerfile that's using pip that even tries to install as non-root. Is it even possible to use it as non-root from a Dockerfile?

I can confirm that using virtual environments inside docker is definitely a thing

pradyunsg commented on news/6409.bugfix.rst:
```diff
@@ -0,0 +1 @@
+This change discourages the usage of pip as root by showing a warning when pip install is run with sudo.
```
```suggestion
Add a warning, discouraging the usage of pip as root, outside a virtual environment.
```

pradyunsg commented on src/pip/_internal/commands/install.py:
```diff
@@ -732,3 +737,25 @@ def create_os_error_message(error, show_traceback, using_user_site):
+
+    On Windows, users may run pip as Administrator without issues.
+    """
+    try:
```
Why is this wrapped in a try?

pradyunsg replied:
See next comment for how I'd suggest rewriting this function.

winsonluk replied:
The try is for `os.getuid()`, which throws an AttributeError when run on Windows. Replied in your other comment.

pradyunsg commented on src/pip/_internal/commands/install.py:
```diff
@@ -732,3 +737,25 @@ def create_os_error_message(error, show_traceback, using_user_site):
+    except AttributeError:
+        warn_sudo = False
+
+    if warn_sudo and not running_under_virtualenv():
```
I'd strongly prefer using early returns, instead of the current structure involving boolean and multiple combinational conditions.

I'd also like to use the `WINDOWS` constant from pip's compat module. 

```py
def warn_if_run_as_root():
    if running_under_virtualenv():
        return
    if WINDOWS:
        return
    if not hasattr(os, "getuid"):
        return
    if os.getuid() != 0:
        return
    logger.warning("...")
```

winsonluk replied:
I like this format. But the reason for the current (admittedly weird) if/elif structure is for [Mypy platform checking](https://mypy.readthedocs.io/en/stable/common_issues.html#python-version-and-system-platform-checks).

Your example would fail Mypy since `if os.getuid() != 0:` will still be run on Mypy's Windows linter, throwing an AttributeError and failing the test suite.

winsonluk replied:
The weird if/elif structure was to follow [Mypy's platform check](https://mypy.readthedocs.io/en/stable/common_issues.html#python-version-and-system-platform-checks) convention and avoid AttributeErrors with `os.getuid`.

But I like this version better.

winsonluk replied:
Update: this still fails Mypy Windows despite the `hasattr` :/

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

src\pip\_internal\cli\req_command.py:177: error: Module has no attribute
"getuid"; maybe "getpid" or "getppid"?
        if os.getuid() != 0:
           ^
Found 1 error in 1 file (checked 140 source files)
```

Reverting to the previous version.

winsonluk replied:
Mypy-compatible version:
```
    if running_under_virtualenv():                                                                  
        return                                                                                      
    if not hasattr(os, "getuid"):                                                                   
        return                                                                                      
    if sys.platform == "win32" or sys.platform == "cygwin":                                         
        return                                                                                      
    elif sys.platform == "darwin" or sys.platform == "linux":                                       
        if os.getuid() != 0:                                                                        
            return                                                                                  
    logger.info("...")
```

pradyunsg commented on src/pip/_internal/commands/install.py:
```diff
@@ -732,3 +737,25 @@ def create_os_error_message(error, show_traceback, using_user_site):
+                        "This could result in the corruption of "
+                        "package directories when a package manager is "
+                        "used. Consider using --user instead.")
+        logger.info(sudo_warning)
```
If it's a warning, it should definitely be a `logger.warning`.

winsonluk replied:
I think this could be either an info or a warning. But I didn't use `logger.warning` because it ends up failing most of the test cases...the tests are run with sudo, but don't like when warnings appear.

Is there an elegant way around this (short of adding exemptions for every test case)?

uranusjr replied:
The most simplistic solution would be to monkeypatch `warn_if_run_as_root` to `lambda: None`.

winsonluk replied:
Changed to logger.warning. No monkeypatch needed.

pradyunsg commented on src/pip/_internal/commands/uninstall.py:
```diff
@@ -3,6 +3,7 @@
 from pip._internal.cli.base_command import Command
 from pip._internal.cli.req_command import SessionCommandMixin
 from pip._internal.cli.status_codes import SUCCESS
+from pip._internal.commands.install import prepare_sudo_warning
```
I'd much rather put this in a different location. I don't want to create a dependency between two commands like this.

Please move the helper function into cli.req_command instead.

pradyunsg commented on src/pip/_internal/commands/install.py:
```diff
@@ -732,3 +737,25 @@ def create_os_error_message(error, show_traceback, using_user_site):
+        warn_sudo = False
+
+    if warn_sudo and not running_under_virtualenv():
+        sudo_warning = ("Running pip as root is not recommended! "
```
I wanna wordsmith on this, so bear with me. I don't think we need to tell users to use --user since pip does do that by default now (if site-packages is not writable). Instead, we should guide users toward using virtual environments.

```
Running pip as root is dangerous. It is possible to break packages installed for your operating system or a different package manager. You should use a virtual environment instead, and run pip as a non-root user.
```

pradyunsg replied:
I also don't think this needs to be a separate variable. Please inline it with the call, and use the black code style here (we're slowly moving toward adopting that across the codebase).

```
logger.warning(
    "..."
    "..."
)
```

winsonluk replied:
Good point on suggesting virtual environments over --user. I've updated the warning.

My only nit is with the last phrase: `You should use a virtual environment instead, and run pip as a non-root user.` It's a bit vague with why and when to use virtual environments. Here's what I suggest:

`Consider using a virtual environment rather than trying to run as root. This may help resolve permission issues.`

pradyunsg reviewed (changes requested):
Alrighty. Consider me convinced that this is a good idea overall.

This is the first time I'm looking at the code in this PR and... I have some opinions about how we should implement this. :)

winsonluk commented:
Changes requested by @pradyunsg have been implemented.

pradyunsg commented on src/pip/_internal/cli/req_command.py:
```diff
@@ -158,6 +160,30 @@ def handle_pip_version_check(self, options):
+        return
+    if sys.platform == "win32" or sys.platform == "cygwin":
+        return
+    elif sys.platform == "darwin" or sys.platform == "linux":
```
This doesn't need to be an elif.

winsonluk replied:
Changed to if

pradyunsg commented on src/pip/_internal/cli/req_command.py:
```diff
@@ -158,6 +160,30 @@ def handle_pip_version_check(self, options):
+    if sys.platform == "win32" or sys.platform == "cygwin":
+        return
+    elif sys.platform == "darwin" or sys.platform == "linux":
+        if os.getuid() != 0:
```
I'd flip the parent conditional and have an early return.

pradyunsg replied:
Unless mypy gets angry. :(

winsonluk replied:
I've tried pretty much every other format...mypy only likes this one :/

pradyunsg commented on src/pip/_internal/cli/req_command.py:
```diff
@@ -158,6 +160,30 @@ def handle_pip_version_check(self, options):
+        return
+    if not hasattr(os, "getuid"):
+        return
+    if sys.platform == "win32" or sys.platform == "cygwin":
```
An inline comment about why we're not using `WINDOWS` here would be helpful:

https://github.com/pypa/pip/blob/a48ad5385b234097d51283b08c3d933fd81ef534/src/pip/_internal/utils/compat.py#L164

winsonluk replied:
```
         return
+    # On Windows, there are no "system managed" Python packages. Installing as
+    # Administrator via pip is the correct way of updating system environments.
     if sys.platform == "win32" or sys.platform == "cygwin":
         return
```

pradyunsg replied:
I think we should also mention that we're not using utils.compat.WINDOWS because of mypy here.

winsonluk replied:
```
    if not hasattr(os, "getuid"):                                                                   
        return                                                                                      
    # On Windows, there are no "system managed" Python packages. Installing as                      
    # Administrator via pip is the correct way of updating system environments.                     
    #                                                                                               
    # We choose sys.platform over utils.compat.WINDOWS here to enable Mypy platform                 
    # checks: https://mypy.readthedocs.io/en/stable/common_issues.html                              
    if sys.platform == "win32" or sys.platform == "cygwin":                                         
        return                                                                                      
    if sys.platform == "darwin" or sys.platform == "linux":                                         
        if os.getuid() != 0:                                                                        
            return 
```

uranusjr commented on src/pip/_internal/cli/req_command.py:
```diff
@@ -158,6 +160,32 @@ def handle_pip_version_check(self, options):
+                   "It is possible to break packages installed for your "
+                   "operating system or by different package managers. "
+                   "You should use a virtual environment rather than trying "
+                   "to run as root. This may help resolve permission issues.")
```
I would try making the message shorter. The longer the warning is, the less likely users would read it.

hroncok replied:
This is shorter, yet OTOH the user has no idea what to do instead:

```python
logger.warning("Running pip as root may destroy your system.")`
```

Do you think providing a link would work? I.e.:

```python
logger.warning("Running pip as root may destroy your system. See https://bit.ly/pip-root for details.")`
```

uranusjr replied:
A link to the documentation feels like a good compromise to me.

winsonluk replied:
There's two situations I can think of where someone would run pip with sudo:

1. Unintentionally: it's often easy to forget which commands need sudo and which don't. In this case, the "Running pip as root may break packages" part of the message should be enough to dissuade them; I don't think we'll need to link an essay.
2. Intentionally: A user may run pip *without* sudo, encounter a permission error, and retry with sudo. We should discourage this by suggesting virtual environments. This is where a link could help imo.

How's this?
```
Running pip as root may break packages and permissions. Permission errors should be fixed with venv: https://docs.python.org/3/library/venv.html 
```

uranusjr replied:
Either of the followings would be a better link IMO:

* https://packaging.python.org/tutorials/installing-packages/#creating-virtual-environments
* https://docs.python.org/3/tutorial/venv.html

The library documentation page covers more on the programmatic interface and isn’t very useful for the audience we target here.

pradyunsg replied:
I like @hroncok's link-based suggestion. I'd say use `https://pip.pypa.io/warnings/root-usage` for that right now. We can setup a redirect in our RTD configuration, to make it point to the right place once this is merged.

It'll also defer the discussion on what the right place is, which I like. 🙃

winsonluk replied:
I still think we should explicitly call out permission errors and venv in the warning message, since that's by far the most common misuse of sudo.

So instead of a generic `https://pip.pypa.io/warnings/root-usage` redirect, how about we specify `https://pip.pypa.io/warnings/venv` and tentatively redirect it to `https://docs.python.org/3/tutorial/venv.html`?

Updated warning message:
`Running pip as root may break packages and permissions. Permission errors should be fixed with venv: https://pip.pypa.io/warnings/venv`

pfmoore replied:
How about "Permission errors should be fixed with venv" -> "If you need to install packages not supplied by your system package manager, use a virtual environment: https://pip.pypa.io/warnings/venv"? Focus on what the user wants to do, rather than simply on fixing the error they got.

pradyunsg replied:
I think we should call this out for what they're doing here, and provide better guidance on what they should be doing there (using a virtualenv, why, what the risks are etc).

winsonluk replied:
Shorter message with a focus on the user:
```
    logger.warning("Running pip as root may break packages and permissions. "                       
                   "You should install packages reliably by using venv: "                                             
                   "https://pip.pypa.io/warnings/venv") 
```

I agree that technical details belong in the link.

pradyunsg reviewed (commented):
Almost there! 💫

Thanks a lot for keeping this flowing @winsonluk! ^.^

winsonluk commented:
I've updated `may break` -> `will break` since site-packages permissions *will* be broken. I've also moved the warning message to *after* the installation logs for more visibility.

winsonluk commented:
Also updating the *other* sudo warning's message to match the phrasing of this one for consistency: "you may want sudo's -H flag" -> "you should use sudo's -H flag"

winsonluk commented:
@pradyunsg @uranusjr Hey folks, just following up on this PR. Any other changes we need, or are we good to merge?

uranusjr reviewed (approved):
This looks good to me, waiting for a second opinion.

pfmoore commented on src/pip/_internal/cli/req_command.py:
```diff
@@ -158,6 +160,33 @@ def handle_pip_version_check(self, options):
+    #
+    # We choose sys.platform over utils.compat.WINDOWS here to enable Mypy platform
+    # checks: https://mypy.readthedocs.io/en/stable/common_issues.html
+    if sys.platform == "win32" or sys.platform == "cygwin":
```
Why are we doing this for cygwin? I'd have imagined that cygwin counts as a "system package manager" for the purposes of a cygwin build of Python. But having said that, I know very little about cygwin and I'm happy to assune it's OK unless cygwin users complain that they don't like it.

uranusjr replied:
It’s been a looong while since I last used Cygwin, but IIRC it does not have a notion of superusers, so it makes sense to skip the checks in it. But it’s probably a good idea to not over-anticipate things and wait for complaints instead, since this is just a line of message anyway.

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

sbidoul commented:
I just stumbled upon on [this part](https://github.com/pypa/pip/blob/8a949a1c52ae47584242e5bb0155d1d1018275ae/src/pip/_internal/cli/base_command.py#L149-L157) of the pip code base which warns about permissions and reckons that the user might be using sudo. I thought this might be relevant to the conversation here.

winsonluk commented:
> I just stumbled upon on [this part](https://github.com/pypa/pip/blob/8a949a1c52ae47584242e5bb0155d1d1018275ae/src/pip/_internal/cli/base_command.py#L149-L157) of the pip code base which warns about permissions and reckons that the user might be using sudo. I thought this might be relevant to the conversation here.

Good call out, this warning often appears when users run as root too. But I think it's useful to keep both cache and root warnings separate, since each case can still exist independently:
- The user may not be running as root, but the cache may still be owned by another user; or,
- The user may be running as root, but the cache and its parent directory are both owned by root.

winsonluk commented:
@uranusjr @pradyunsg @pfmoore Rebase is complete. We still need to set up a redirect from https://pip.pypa.io/warnings/venv to https://docs.python.org/3/tutorial/venv.html before merging, but that should be the last step!

winsonluk commented:
#9680 configures the redirect. would appreciate a review :)

pradyunsg commented:
> [pip.pypa.io/warnings/venv](https://pip.pypa.io/warnings/venv) to [docs.python.org/3/tutorial/venv.html](https://docs.python.org/3/tutorial/venv.html)

Configured the redirect.

winsonluk commented:
Thanks @pradyunsg, this should be good to merge then.

pradyunsg reviewed (commented):
Could you do a rebase of this PR? I'd like to make sure it still passes the current linter setup.

winsonluk commented:
> Could you do a rebase of this PR? I'd like to make sure it still passes the current linter setup.

@pradyunsg Sure, rebased without issues.

pradyunsg commented:
Thanks for this PR! ^.^

graue70 commented:
I'm adding the current version of the warning message so this PR with some explanations can be found more easily:

`WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv`

graue70 commented:
> I think we should. Even in docker, installing with pip as root can lead to serious problems.

> Using root in a Docker container is ultimately still not best practice, and it’s arguably still not a bad idea to show the warning in this situation.

> I can confirm that using virtual environments inside docker is definitely a thing

Both the [documentation of the official python docker images](https://hub.docker.com/_/python/) and the [official docker docs](https://docs.docker.com/language/python/build-images/#create-a-dockerfile-for-python) install pip packages with sudo. I can't find any information on the web as to how it should be done instead. Would you please explain how I'm supposed to get rid of the warning when using a `Dockerfile`? @hroncok @uranusjr @xavfernandez

uranusjr commented:
[Processes In Containers Should Not Run As Root](https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b)

[Deploying Python Applications with Docker - A Suggestion](https://glyph.twistedmatrix.com/2015/03/docker-deploy-double-dutch.html) (It covers a ton of stuff, but your can search for *virtualenv* to find the relevant parts.)

Maybe we should write a documentation page on this, and point `https://pip.pypa.io/warnings/venv` to that page instead.
